### PR TITLE
Support `{prefix}` and `{lowerprefix}` markers in `config.json` `dl` key

### PR DESCRIPTION
**Pull request by drmikehenry**
Hello,

The crates.io-index Git repository uses a nice directory structure to keep individual directory sizes under control.

When mirroring crates.io, it's useful to store crate files in a similar directory structure for the same reasons.

Cargo provides "markers" for use in the `dl` key of the `config.json` file in crates.io-index to allow flexibility in mapping a crate's name and version into a URL for the crate.  The marker `{crate}` is replaced by the crate's name, and the marker `{version}` is replaced with the crate's version.  The default URL template is `https://crates.io/api/v1/crates/{crate}/{version}/download`.

Currently, if a mirror of crates.io stores crates in a directory structure similar to that of crates.io-index, it's up to the server to construct the directory name from the crate name.  This eliminates trivial web servers and `file:` URLs from hosting such a tree of crates.

This pull requests adds two new markers for the `dl` key in `config.json`, allowing Cargo to supply the directory name as part of the URL.  The marker `{lowerprefix}` is the same directory name used within crates.io-index; it is calculated from the crate name converted to lowercase.  The marker `{prefix}` is similar, but it uses the crate name as-is (without case conversion), which is useful for supporting older versions of Cargo that lack these markers; for example, nginx rewrite rules can easily construct `{prefix}` but can't perform case-conversion to construct `{lowerprefix}`.  These new markers will provide implementation flexibility and simplicity for crate mirror servers.

### CODE REVIEW

**Comment by alexcrichton** [comment]
Thanks for the PR and seems reasonable to me! I'm not sure that `{prefix}` and `{lowerprefix}` are the best names here but we can always bikeshed. In any case this is some new stable surface area, so I'm going to canvas the team:

@rfcbot fcp merge

**Comment by joshtriplett** [comment]
I've had several discussions with some of the folks involved with creating the original directory structure with these prefix subdirectories, and each time, they've expressed regret at the use of those directories; they're much less relevant on modern filesystems. That shouldn't stop us from supporting the directory structure we have; I'm only mentioning it because anyone creating a *new* registry should seriously evaluate the directory structure and figure out what works best, rather than assuming that "Rust's upstream registry did it so it's probably a good idea". :smile:

That said, having support for this seems reasonable:
@rfcbot reviewed

**Comment by drmikehenry** [comment]
Thanks for considering this pull request.  As always (for me, anyway :-)), naming is hard.  I'm happy to change `{prefix}` and `{lowerprefix}` to whatever is preferred.

File: src/doc/src/reference/registries.md [hunk 1]
```diff
@@ -128,8 +128,11 @@ looks like:
 The keys are:
 - `dl`: This is the URL for downloading crates listed in the index. The value
   may have the markers `{crate}` and `{version}` which are replaced with the
-  name and version of the crate to download. If the markers are not present,
-  then the value `/{crate}/{version}/download` is appended to the end.
+  name and version of the crate to download, or the marker `{prefix}` which is
+  replaced with the crate's prefix, or the marker `{lowerprefix}` which is
+  replaced with the crate's prefix converted to lowercase.  If none of the
+  markers are present, then the value `/{crate}/{version}/download` is appended
+  to the end.  See below for more about crate prefixes.
```
[the remark this answers is not in the document]
Comment (drmikehenry) [reply src/doc/src/reference/registries.md]: Thanks; I've incorporated your wording improvements into the pull request.

**Pull request merged**